controllers/seedgen_controller.go

```diff
@@ -0,0 +1,635 @@
+		if client.IgnoreNotFound(err) != nil {
+			r.Log.Info("Error when checking namespaces: %w", err)
+		}
+		return
+	}
+
+	re := regexp.MustCompile(`^(open-cluster-management-agent|open-cluster-management-addon)`)
+	for _, ns := range namespaces.Items {
+		if re.MatchString(ns.ObjectMeta.Name) {
+			acmNsList = append(acmNsList, ns.ObjectMeta.Name)
+		}
+	}
+	return
+}
+
+// Clean up ACM and other resources on the cluster
+func (r *SeedGeneratorReconciler) cleanupClusterResources(ctx context.Context, hubClient client.Client) error {
+	if !r.managedClusterExists(ctx, hubClient) {
+		r.Log.Info("ManagedCluster does not exist on hub. Skipping cleanup")
+		return nil
+	}
+
+	// Ensure that the dependent resources are deleted
+	deleteOpts := []client.DeleteOption{
+		client.PropagationPolicy(metav1.DeletePropagationForeground),
+	}
+
+	// Deregister from ACM on the hub
+	managedcluster := &clusterv1.ManagedCluster{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: clusterName,
+		}}
+	if err := hubClient.Delete(ctx, managedcluster, deleteOpts...); client.IgnoreNotFound(err) != nil {
+		return fmt.Errorf("Failed to delete managedcluster from hub: %w", err)
+	}
+
+	// TODO: For some reason, the managedcluster deletion is returning immediately, rather than
+	// blocking while the deletion occurs in the foreground. Maybe because it's deleting on the hub?
+	// As a workaround, we'll poll until the cluster is deleted.
+	interval := 10 * time.Second
+	maxRetries := 90 // ~15 minutes
+	current := 0
+	r.Log.Info("Waiting until managedcluster is deleted")
+	for r.managedClusterExists(ctx, hubClient) {
+		if current < maxRetries {
+			time.Sleep(interval)
+			current += 1
+		} else {
+			return fmt.Errorf("Timed out waiting for managedcluster deletion")
+		}
+	}
+
+	// Trigger deletion for any remaining ACM namespaces
+	for _, nsName := range r.currentAcmNamespaces(ctx) {
+		ns := &corev1.Namespace{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: nsName,
+			}}
+		if err := r.Client.Delete(ctx, ns, deleteOpts...); client.IgnoreNotFound(err) != nil {
+			return fmt.Errorf("Failed to delete namespace %s: %w", nsName, err)
+		}
+	}
+
+	// Verify ACM namespaces have been deleted
+	current = 0
+	r.Log.Info("Waiting until ACM namespaces are deleted")
+	for len(r.currentAcmNamespaces(ctx)) > 0 {
+		if current < maxRetries {
+			time.Sleep(interval)
+			current += 1
+		} else {
+			return fmt.Errorf("Timed out waiting for ACM namespace deletion")
+		}
+	}
+
+	// Delete remaining cluster resources leftover from ACM (or install)
+
+	ns := &corev1.Namespace{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "assisted-installer",
+		}}
+	if err := r.Client.Delete(ctx, ns, deleteOpts...); client.IgnoreNotFound(err) != nil {
+		return fmt.Errorf("Failed to delete assisted-installer namespace: %w", err)
+	}
+
+	// TODO: We could do a query similar to the namespace query above to look for all remaining open-cluster-management.io CRDs
+	crds := []string{
```

Comment:
These as well ? 
clusterclaims.cluster.open-cluster-management.io
klusterlets.operator.open-cluster-management.io

Reply:
Those get automatically cleaned up as part of deregistration, so I've never needed to manually delete them. I could change this to a query similar to the namespace one, searching for CRDs matching *.open-cluster-management.io and delete whatever's there